File: Tree Algorithms/subtree_queries.cpp
#include <bits/stdc++.h>
using namespace std;

int t = 1;

void dfs(int node, int p, vector<vector<int>> &adj, vector<long long> &value,
         vector<int> &start, vector<int> &end, vector<long long> &lt)
{
    start[node] = t;
    lt[t++] = value[node];

    for (int child : adj[node])
    {
        if (child != p)
        {
            dfs(child, node, adj, value, start, end, lt);
        }
    }

    end[node] = t;
    lt[t++] = value[node];
}

void update(int x, long delta, vector<long long> &BIT){
    for(++x; x < BIT.size(); x += (-x & x)){
        BIT[x] += delta;
    }
}

long long querySum(int x, vector<long long> &BIT){
    long sum = 0;
    for(++x; x > 0; x -= (-x & x)){
        sum += BIT[x];
    }
    return sum;
}

long long rangeSum(int l, int r, vector<long long> &BIT){
    return querySum(r, BIT) - querySum(l-1, BIT);
}


int main()
{
    std::ios::sync_with_stdio(false);
    int n;
    int q;
    cin >> n;
    cin >> q;
    vector<long long> value(n + 1, 0);
    for (int i = 1; i < n + 1; i++)
    {
        cin>>value[i];
    }

    vector<int> start(n + 1, 0);
    vector<int> end(n + 1, 0);
    vector<long long> lt(2 * n + 1, 0);
    vector<long long> bit(2 * n + 2, 0);
    vector<vector<int>> adj(n+1, vector<int>());
    int u;
    int v;
    for (int i = 1; i < n; i++)
    {
        cin >> u;
        cin >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    dfs(1, -1, adj, value, start, end, lt);
    for(int i = 1; i < t; i++){
        update(i, lt[i], bit);
    }
    while(q --> 0){
        int op;
        int node;
        cin >> op;
        cin >> node;
        if(op == 1){
            long long val;
            cin >> val;
            long long delta = val - value[node];
            update(start[node], delta, bit);
            update(end[node], delta, bit);
            value[node] = val;
        }
        else if(op == 2){
            long long ans = rangeSum(start[node], end[node], bit);
            ans = ans/2;
            cout<< ans << endl;
        }
    }
    return 0;
}

File: Tree Algorithms/SubtreeQueries.java
import java.util.*;
import java.io.*;

public class SubtreeQueries{
    static int time = 1;
    static int[] start;
    static int[] end;
    static ArrayList<Integer>[] adj;
    static long[] lt;
    static int[] value;

    static long[] BIT;

    static void update(int x, long delta){
        ++x;
        while(x < BIT.length){
            BIT[x] += delta;
            x += (-x & x);
        }
    }

    static long querySum(int x){
        long sum = 0;
        ++x;
        while(x > 0){
            sum += BIT[x];
            x -= (-x & x);
        }
        return sum;
    }
    
    static class FastReader{
        BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st=new StringTokenizer("");
		String next() {
			while (!st.hasMoreTokens())
				try { 
                                        st=new StringTokenizer(br.readLine());				               
                                } catch (IOException e) {}
			return st.nextToken();
		}
		
		int nextInt() {
			return Integer.parseInt(next());
		}
		long nextLong() {
			return Long.parseLong(next());
		}
    }
    public static void main(String[] args) {
        PrintWriter out=new PrintWriter(System.out);
        FastReader in = new FastReader();
        
        int n = in.nextInt();
        int q = in.nextInt();
        value = new int[n+1];
        for(int i = 1; i < n+1; i++){
            value[i] = in.nextInt();
        }

        start = new int[n+1];
        end = new int[n+1];
        lt = new long[2 * n + 5];
        BIT = new long[lt.length];
        
        
        adj = new ArrayList[n+1];
        for(int i = 0; i < n+1; i++){
            adj[i] = new ArrayList<>();
        }
        int u;
        int v;
        for(int i = 1; i < n; i++){
            u = in.nextInt();
            v = in.nextInt();
            adj[u].add(v);
            adj[v].add(u);
        }
        dfs(1, -1);
        
        for(int i = 1; i < time; i++){
            update(i, lt[i]);
        }

        while(q --> 0){
            int op = in.nextInt();
            int node = in.nextInt();
            if(op == 1){
                int val = in.nextInt();
                long delta = val - value[node];
                update(start[node], delta);
                update(end[node], delta);
                value[node] = val;
            }
            else if(op == 2){
                long ans = querySum(end[node]) - querySum(start[node]-1);
                ans = ans/2;
                out.println(ans);
            }
        }
        out.close();
    }

    static void dfs(int node, int p){
        start[node] = time;
        lt[time++] = value[node];

        for(int child : adj[node]){
            if(child != p){
                dfs(child, node);
            }
        }

        end[node] = time;
        lt[time++] = value[node];
    }
}

